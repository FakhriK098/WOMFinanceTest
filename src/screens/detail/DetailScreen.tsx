import React, { useEffect } from 'react';
import {
  ActivityIndicator,
  Image,
  ScrollView,
  StyleSheet,
  View,
  useColorScheme,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation, useRoute, RouteProp } from '@react-navigation/native';
import {
  Text,
  Chip,
  SpriteGrid,
  EvolutionList,
  InfoRow,
  Section,
} from '@components';
import { RootState, AppDispatch } from '@store';
import {
  fetchPokemonDetail,
  clearDetail,
} from '@store/slices/pokemonDetailSlice';
import { colors } from '@theme/colors';
import { spacing } from '@theme/spacing';
import { typography } from '@theme/typography';
import type { RootStackParamList } from '@typings/navigation';

const DetailScreen = () => {
  const dispatch = useDispatch<AppDispatch>();
  const route = useRoute<RouteProp<RootStackParamList, 'Detail'>>();
  const navigation = useNavigation();
  const { data, loading, error } = useSelector(
    (s: RootState) => s.pokemonDetail,
  );
  const isDark = useColorScheme() === 'dark';
  const id = route.params?.id;

  useEffect(() => {
    if (id) dispatch(fetchPokemonDetail(id));
    return () => {
      dispatch(clearDetail());
    };
  }, [dispatch, id]);

  useEffect(() => {
    navigation.setOptions({
      title: data?.name ? data.name : 'Detail',
      headerStyle: {
        backgroundColor: isDark
          ? colors.background.dark
          : colors.background.light,
      },
      headerTintColor: isDark ? colors.text.dark : colors.text.light,
    });
  }, [navigation, data?.name, isDark]);

  if (loading && !data) {
    return (
      <View
        style={[
          styles.center,
          {
            backgroundColor: isDark
              ? colors.background.dark
              : colors.background.light,
          },
        ]}
      >
        <ActivityIndicator />
      </View>
    );
  }

  if (error) {
    return (
      <View
        style={[
          styles.center,
          {
            backgroundColor: isDark
              ? colors.background.dark
              : colors.background.light,
          },
        ]}
      >
        <Text
          size={typography.sm}
          color={isDark ? colors.text.mutedDark : colors.text.mutedLight}
        >
          {error}
        </Text>
      </View>
    );
  }

  if (!data) {
    return (
      <View
        style={[
          styles.center,
          {
            backgroundColor: isDark
              ? colors.background.dark
              : colors.background.light,
          },
        ]}
      >
        <Text
          size={typography.sm}
          color={isDark ? colors.text.mutedDark : colors.text.mutedLight}
        >
          Tidak ada data
        </Text>
      </View>
    );
  }

  return (
    <ScrollView
      style={[
        styles.container,
        {
          backgroundColor: isDark
            ? colors.background.dark
            : colors.background.light,
        },
      ]}
      contentContainerStyle={styles.content}
      showsVerticalScrollIndicator={false}
    >
      {!!data.image && (
        <Image
          source={{ uri: data.image }}
          style={[styles.hero, { borderColor: data.color }]}
          resizeMode="contain"
        />
      )}
      <View style={styles.section}>
        <Text
          weight="bold"
          size={typography.xxl}
          color={isDark ? colors.text.dark : colors.text.light}
        >
          {data.name}
        </Text>
        <View style={styles.rowWrap}>
          {data.types.map(t => (
            <Chip
              key={t}
              label={t}
              color={data.color}
              textColor={colors.white}
            />
          ))}
        </View>
      </View>

      <Section title="Informasi" isDark={isDark}>
        <InfoRow label="Tinggi" value={data.height} isDark={isDark} />
        <InfoRow label="Berat" value={data.weight} isDark={isDark} />
        <InfoRow label="Base XP" value={data.base_experience} isDark={isDark} />
      </Section>

      <Section title="Abilities" isDark={isDark}>
        <View style={styles.rowWrap}>
          {data.abilities.map(a => (
            <Chip
              key={a}
              label={a}
              color={data.color}
              textColor={colors.white}
            />
          ))}
        </View>
      </Section>

      {data.sprites.length > 0 && (
        <Section title="Sprites" isDark={isDark}>
          <SpriteGrid sprites={data.sprites} />
        </Section>
      )}

      {data.evolutions.length > 0 && (
        <Section title="Evolutions" isDark={isDark}>
          <EvolutionList
            items={data.evolutions}
            onPressItem={nid => dispatch(fetchPokemonDetail(nid))}
          />
        </Section>
      )}
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1 },
  content: { padding: spacing.lg },
  center: { flex: 1, alignItems: 'center', justifyContent: 'center' },
  hero: {
    width: '100%',
    height: 180,
    borderWidth: 2,
    borderRadius: 12,
    marginBottom: spacing.lg,
  },
  section: { marginBottom: spacing.xl },
  rowWrap: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    gap: spacing.sm,
    marginTop: spacing.sm,
  },
});

export default DetailScreen;
