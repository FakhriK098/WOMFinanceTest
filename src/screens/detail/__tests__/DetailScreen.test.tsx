import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { useSelector, useDispatch } from 'react-redux';
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
import * as RN from 'react-native';

// Mock navigation + route
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ setOptions: jest.fn() }),
  useRoute: () => ({ params: { id: 1 } }),
}));
jest.mock('@store', () => ({ store: {} }));
jest.mock('@store/slices/pokemonDetailSlice', () => ({
  fetchPokemonDetail: () => ({ type: 'fetchDetail' }),
  clearDetail: () => ({ type: 'clearDetail' }),
}));

import DetailScreen from '../DetailScreen';

describe('DetailScreen', () => {
  const useSelectorSpy = useSelector as unknown as jest.Mock;
  const useDispatchSpy = useDispatch as unknown as jest.Mock;

  beforeEach(() => {
    jest.spyOn(RN, 'useColorScheme').mockReturnValue('light');
    useDispatchSpy.mockReturnValue(jest.fn());
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('shows loading state (snapshot)', () => {
    useSelectorSpy.mockImplementation((sel: any) =>
      sel({ pokemonDetail: { data: null, loading: true, error: null } }),
    );
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<DetailScreen />);
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('shows error state (snapshot)', () => {
    useSelectorSpy.mockImplementation((sel: any) =>
      sel({ pokemonDetail: { data: null, loading: false, error: 'Error' } }),
    );
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<DetailScreen />);
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    act(() => tr!.unmount());
  });

  it('renders detail content (snapshot)', () => {
    const data = {
      id: 1,
      name: 'bulbasaur',
      color: '#22c55e',
      image: 'http://img',
      types: ['grass', 'poison'],
      height: 7,
      weight: 69,
      base_experience: 64,
      abilities: ['overgrow', 'chlorophyll'],
      sprites: ['http://s1', 'http://s2'],
      evolutions: [
        { id: 2, name: 'ivysaur', image: 'http://img2' },
        { id: 3, name: 'venusaur', image: 'http://img3' },
      ],
    };
    useSelectorSpy.mockImplementation((sel: any) =>
      sel({ pokemonDetail: { data, loading: false, error: null } }),
    );
    let tr: renderer.ReactTestRenderer;
    act(() => {
      tr = renderer.create(<DetailScreen />);
    });
    expect(tr!.toJSON()).toMatchSnapshot();
    const nameNode = tr!.root.find(
      (inst) => inst.props && inst.props.children === 'bulbasaur',
    );
    expect(nameNode).toBeTruthy();
    act(() => tr!.unmount());
  });
});
