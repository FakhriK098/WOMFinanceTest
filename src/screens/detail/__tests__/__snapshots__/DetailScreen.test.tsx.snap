// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailScreen renders detail content (snapshot) 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "padding": 20,
    }
  }
  showsVerticalScrollIndicator={false}
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <View>
    <Image
      resizeMode="contain"
      source={
        {
          "uri": "http://img",
        }
      }
      style={
        [
          {
            "borderRadius": 12,
            "borderWidth": 2,
            "height": 180,
            "marginBottom": 20,
            "width": "100%",
          },
          {
            "borderColor": "#22c55e",
          },
        ]
      }
    />
    <View
      style={
        {
          "marginBottom": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#111827",
              "fontSize": 24,
              "fontWeight": "700",
            },
            undefined,
          ]
        }
      >
        bulbasaur
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "marginTop": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "alignSelf": "flex-start",
                "borderRadius": 12,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              {
                "backgroundColor": "#22c55e",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 12,
                  "fontWeight": "400",
                },
                undefined,
              ]
            }
          >
            grass
          </Text>
        </View>
        <View
          style={
            [
              {
                "alignSelf": "flex-start",
                "borderRadius": 12,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              {
                "backgroundColor": "#22c55e",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 12,
                  "fontWeight": "400",
                },
                undefined,
              ]
            }
          >
            poison
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#111827",
              "fontSize": 18,
              "fontWeight": "700",
            },
            undefined,
          ]
        }
      >
        Informasi
      </Text>
      <View
        style={
          {
            "height": 8,
          }
        }
      />
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingVertical": 6,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#6B7280",
                "fontSize": 14,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          Tinggi
        </Text>
        <Text
          style={
            [
              {
                "color": "#111827",
                "fontSize": 14,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          7
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingVertical": 6,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#6B7280",
                "fontSize": 14,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          Berat
        </Text>
        <Text
          style={
            [
              {
                "color": "#111827",
                "fontSize": 14,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          69
        </Text>
      </View>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingVertical": 6,
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#6B7280",
                "fontSize": 14,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          Base XP
        </Text>
        <Text
          style={
            [
              {
                "color": "#111827",
                "fontSize": 14,
                "fontWeight": "400",
              },
              undefined,
            ]
          }
        >
          64
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#111827",
              "fontSize": 18,
              "fontWeight": "700",
            },
            undefined,
          ]
        }
      >
        Abilities
      </Text>
      <View
        style={
          {
            "height": 8,
          }
        }
      />
      <View
        style={
          {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "marginTop": 8,
          }
        }
      >
        <View
          style={
            [
              {
                "alignSelf": "flex-start",
                "borderRadius": 12,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              {
                "backgroundColor": "#22c55e",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 12,
                  "fontWeight": "400",
                },
                undefined,
              ]
            }
          >
            overgrow
          </Text>
        </View>
        <View
          style={
            [
              {
                "alignSelf": "flex-start",
                "borderRadius": 12,
                "paddingHorizontal": 8,
                "paddingVertical": 4,
              },
              {
                "backgroundColor": "#22c55e",
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontSize": 12,
                  "fontWeight": "400",
                },
                undefined,
              ]
            }
          >
            chlorophyll
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#111827",
              "fontSize": 18,
              "fontWeight": "700",
            },
            undefined,
          ]
        }
      >
        Sprites
      </Text>
      <View
        style={
          {
            "height": 8,
          }
        }
      />
      <View
        style={
          {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "uri": "http://s1",
            }
          }
          style={
            {
              "height": 72,
              "width": 72,
            }
          }
        />
        <Image
          resizeMode="contain"
          source={
            {
              "uri": "http://s2",
            }
          }
          style={
            {
              "height": 72,
              "width": 72,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 24,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#111827",
              "fontSize": 18,
              "fontWeight": "700",
            },
            undefined,
          ]
        }
      >
        Evolutions
      </Text>
      <View
        style={
          {
            "height": 8,
          }
        }
      />
      <RCTScrollView
        contentContainerStyle={
          {
            "gap": 8,
          }
        }
        data={
          [
            {
              "id": 2,
              "image": "http://img2",
              "name": "ivysaur",
            },
            {
              "id": 3,
              "image": "http://img3",
              "name": "venusaur",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        showsHorizontalScrollIndicator={false}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              [
                {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "borderRadius": 12,
                  "padding": 8,
                  "width": 100,
                }
              }
            >
              <Image
                resizeMode="contain"
                source={
                  {
                    "uri": "http://img2",
                  }
                }
                style={
                  {
                    "height": 72,
                    "marginBottom": 4,
                    "width": 72,
                  }
                }
              />
              <Text
                style={
                  [
                    {
                      "color": "#111827",
                      "fontSize": 12,
                      "fontWeight": "400",
                    },
                    undefined,
                  ]
                }
              >
                ivysaur
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={
              [
                {
                  "flexDirection": "row",
                },
                null,
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "borderRadius": 12,
                  "padding": 8,
                  "width": 100,
                }
              }
            >
              <Image
                resizeMode="contain"
                source={
                  {
                    "uri": "http://img3",
                  }
                }
                style={
                  {
                    "height": 72,
                    "marginBottom": 4,
                    "width": 72,
                  }
                }
              />
              <Text
                style={
                  [
                    {
                      "color": "#111827",
                      "fontSize": 12,
                      "fontWeight": "400",
                    },
                    undefined,
                  ]
                }
              >
                venusaur
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RCTScrollView>
`;

exports[`DetailScreen shows error state (snapshot) 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#6B7280",
          "fontSize": 14,
          "fontWeight": "400",
        },
        undefined,
      ]
    }
  >
    Error
  </Text>
</View>
`;

exports[`DetailScreen shows loading state (snapshot) 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <ActivityIndicator />
</View>
`;
